File: src/python/problems/count_nodes_test.py
# https://leetcode.com/problems/count-complete-tree-nodes

from typing import Optional
from core.problem_base import *
from models.tree_node import TreeNode

class CountNodes(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> int:
        def Count(root):
            if not root:
                return 0

            node = root
            leftDepth = 0
            while node:
                node = node.left
                leftDepth += 1

            node = root
            rightDepth = 0
            while node:
                node = node.right
                rightDepth += 1

            if leftDepth == rightDepth:
                return 2 ** leftDepth - 1

            return 1 + Count(root.left) + Count(root.right)

        return Count(root)

if __name__ == '__main__':
    TestGen(CountNodes) \
        .Add(lambda tc: tc.Param([1,2,3,4,5,6]).Result(6)) \
        .Add(lambda tc: tc.Param([]).Result(0)) \
        .Add(lambda tc: tc.Param([1]).Result(1)) \
        .Run()

File: src/go/problems/count_nodes_test.go
// https://leetcode.com/problems/count-complete-tree-nodes/

package problems

import (
	"math"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type CountNodes struct{}

func TestCountNodes(t *testing.T) {
	gen := core.TestSuite[CountNodes]{}
	gen.Add(func(tc *core.TestCase) {
		root := &TreeNode{
			Val:   1,
			Left:  &TreeNode{Val: 2, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 5}},
			Right: &TreeNode{Val: 3, Left: &TreeNode{Val: 6}},
		}
		tc.Param(root).Result(6)
	}).Run(t)
}

func (CountNodes) Solution(root *TreeNode) int {
	var count func(root *TreeNode) int
	count = func(root *TreeNode) int {
		if root == nil {
			return 0
		}

		leftLevel := 0.
		node := root
		for node != nil {
			node = node.Left
			leftLevel++
		}

		rightLevel := 0.
		node = root
		for node != nil {
			node = node.Right
			rightLevel++
		}

		if leftLevel == rightLevel {
			return int(math.Pow(2.0, leftLevel)) - 1
		}

		return 1 + count(root.Left) + count(root.Right)
	}

	return count(root)
}
